### DISCUSSION TITLE
Best practice for grouped allreduce?

### DISCUSSION THREAD

**Discussion by zhuzilin** [horovod/horovod#2773] [Q&A]
**Environment:**
PyTorch 1.7.0 + Horovod 0.21

**Your question:**
It's so nice that we got grouped allreduce in version 0.21 and our team are trying to use it to create better fusion policy with deterministic order as mentioned in the [official article](https://eng.uber.com/horovod-v0-21/). We've discovered the basic rule for this functionality and hope you could check if this is correct:
- When non-zero `num_groups` is set, the tensors in the same group will enter the `DoGroupedAllReduce` together only if all gradients of them are ready.
- If user sets the environment variable `HOROVOD_DISABLE_GROUP_FUSION`, tensors in different groups are not allowed to be fused together. Otherwise, the C++ backend will simply ignore the `group_id` on those tensors. (Functions like `GetGroupTensorNames` are only called in the scope where `state.disable_group_fusion` is true).
- Tensors in the same group may still be fused into many `response`s. To make the tensors in a group fused into one `response`, we need to set `HOROVOD_DISABLE_GROUP_FUSION` and a very large `HOROVOD_FUSION_THRESHOLD`, which is kind of hacky.

If our understanding is correct, we think that it will be more intuitive if we could allow the user to fusion the whole group by some options instead of the hacky way we mentioned above. The current setting will blend the deterministic fusion with the origin dynamic fusion, which may be a little confusing.

Also, we hope you could give us more freedom by providing interfaces like #2523. The current implementation will first sort the `named_parameters` by their name and split them evenly. However, we often need to fuse the parameters that are used around the same time so that they won't wait each other for too long. And because the communication of the last fusion group could not overlap with the backward computation, we often set the last group a little smaller than others. Those details may help improve the scaling efficiency and the overall training speed a lot in practice and to implement them, the user needs to be able to control the order of the parameters and the size of the group by themselves. Therefore, we hope that horovod could have such interface for a even better performance.

Anyway, thank you so much for giving us this nice grouped allreduce functionality! We'd love to discussion with you more about it and if you have any interest in the points we mentioned above, we'd love to contribute!

Thank you for your time on this long issue and wish you happy New Year :).

Gently ping @romerojosh @tgaddair

**Comment by romerojosh** [comment]
@zhuzilin Excellent question and summary of how the grouped allreduce mechanism works! The three points you make are correct. I wanted to comment on your third point:
> Tensors in the same group may still be fused into many responses. To make the tensors in a group fused into one response, we need to set HOROVOD_DISABLE_GROUP_FUSION and a very large HOROVOD_FUSION_THRESHOLD, which is kind of hacky.
As it currently stands, the requirements to guarantee all the gradient tensors fused into a single group are:
1. Set `num_groups=1` in `DistributedOptimizer`/`DistributedGradientTape`.
2. Set `HOROVOD_FUSION_THRESHOLD` to a value large enough to contain all the gradient tensors in the single group.
3. Set `HOROVOD_DISABLE_GROUP_FUSION=1` so no other tensor allreduces (e.g. collectives calls outside the optimizer) are fused into the group, potentially making the fused group larger than `HOROVOD_FUSION_THRESHOLD` bytes.

I agree that we could make achieving this a bit more straightforward by automatically resizing the fusion buffer to be able to fit the largest tensor group, thus removing the need for the user to manually specify a large enough `HOROVOD_FUSION_THRESHOLD`. What do you think @tgaddair?

On #2523, I think having this type of control over the group formation would be a great addition and would provide users a lot more opportunities for fine-tuned optimization of the communication schedule. In fact, at some point, I experimented with a version of the grouped allreduce that leveraged the existing autotuner in Horovod to try and tune the groups, but ended up dropping the idea as it was a bit too complicated for the initial PR. This PR could be a good step towards revisiting that idea.  Apologies for the delay in reviewing but will try and take a look as soon as I can.

**Comment by zhuzilin** [comment]
@romerojosh
Thank you for your reply! 
> In fact, at some point, I experimented with a version of the grouped allreduce that leveraged the existing autotuner in Horovod to try and tune the groups, but ended up dropping the idea as it was a bit too complicated for the initial PR.

In our experience, it's hard to have a general policy as the bottleneck differs among models. I think it is more aproachable to provide some choices and leave the fine-tuning api to the user.

**Comment by tgaddair** [comment]
> I agree that we could make achieving this a bit more straightforward by automatically resizing the fusion buffer to be able to fit the largest tensor group, thus removing the need for the user to manually specify a large enough HOROVOD_FUSION_THRESHOLD.

I think this is a great idea @romerojosh!  Do you have bandwidth to put together a PR for this?

**Comment by romerojosh** [comment]
@tgaddair  I'm not sure I will be able to get to this soon, but can put it on my list. One question would be when this dynamic fusion buffer resizing should be enabled? I don't think we want it to be dynamically resized in the general case since the some tensor fusion/cycle time combinations could create very large buffers and lead to memory issues for users. Same would apply when using groups, but still allowing fusion. We could potentially only enable it automatically if groups are used and `DISABLE_GROUP_FUSION` is set, but that is a bit opaque. Another option would be to add yet another knob to enable dynamic fusion sizing, but not sure that is quite right either. Any thoughts on that?

**Comment by tgaddair** [comment]
@romerojosh good question.  I think for now an additional knob is the best I can think of.  As long as we set reasonable defaults, though, I think this okay.

**Comment by GHGmc2** [comment]
@romerojosh For "I agree that we could make achieving this a bit more straightforward by automatically resizing the fusion buffer to be able to fit the largest tensor group, thus removing the need for the user to manually specify a large enough HOROVOD_FUSION_THRESHOLD."

May I know if it's available in Horovod now? What if a user set `num_group=1` and gradient size is larger than fusion buffer?
